New signup form changes

kamfonik opened the pull request:
This should be mostly identical to #16 except the commits are on top of #19.

Original comment from #16:
This updates addusers.py to work with a new version of the signup form that gathers more info. This includes the following code changes:

    form data is now stored in User and Project classes instead of dictionaries
    more distinct scenarios are handled, which are specified explicitly in the form entry:
        new user + new project
        existing user + new project
        new user + existing project
        existing user + existing project
    the script skips entries that are not yet approved or for which a Helpdesk ticket is not yet open.
    A separate notification email is sent for new user accounts and new projects.

If anyone wants a link to my testing form/spreadsheet to help with the code review, let me know and I'll send it via email - I don't want to post it publicly.

Don't worry about pep8/code style too much; the next PR to this repo will be a pep8 overhaul of the 3 files excluded from flake8 checks. I want to get this one merged before wasting time pep8-izing code that would be removed.

kamfonik commented:
@robbaronbu do you have time to take a quick look at this?  It should be pretty much the same as #16 which you approved, most of the commits are the same (with minor changes to fix merge conflicts).   

The last 3 commits are new since your approval.

kamfonik commented:
@xuhang57 yeah, we definitely need testing in this repo.  Once we get the basics in place for the helpdesk to take over using this script, we can work on testing.

#17 is a companion to this PR - it is marked WIP because we have to test that the email it generates against the helpdesk ticket system, but there is a meeting for that this afternoon.

xuhang57 commented:
@kamfonik  I will take a look at that PR once you guys finish the meeting. 

And yes, I just want to make sure to give a careful/validate review. looking at the code is not so effective as actually testing the coed 👍 . Thanks for considering it and looking forward to it.

kamfonik commented:
Of course, git doesn't know it resolves the merge conflict.

kamfonik commented:
While testing the merge conflicts I discovered some bugs in the way the script was marking rows for copying/skipping, and realized there was a need to differentiate between the requesting user and their additional users.
  
The additional commit addresses most of the issues but I have also opened  #21, which is not urgent but we should fix it later.  Fixing it may require changes to the flow of information through the script.

kamfonik commented on addusers.py:
```diff
@@ -388,11 +389,22 @@ def parse_rows(rows):
                     response = openstack.grant_role(auth_url, ks_project.id, ks_user.id, ks_member_role.id)
-                copy_index.append(user.row)
+                
+                if user.is_requestor: 
```
This stops the script  from repeatedly marking the same row for deletion, which will cause neighboring rows to be deleted by accident.  This is the quick fix; I don't really like it.  But it works for now and hopefully a more elegant solution will come with the fix to #21

kamfonik commented on addusers.py:
```diff
@@ -388,11 +389,22 @@ def parse_rows(rows):
+                    if user.is_requestor:
+                        raise ItemNotFoundError('User', user.name)
+                    else:
+                        # We don't treat this as a critical error
```
We care whether the person who filled out the form said they had an existing account, but doesn't.  

We don't actually care so much if someone they listed as an additional user doesn't exist.  That's much easier to fix later, so we just output a notice that we are skipping that person.

kamfonik commented:
@xuhang57 could you take a quick look at the commits added since you approved, and just check for anything outrageous.

xuhang57 commented on addusers.py:
```diff
@@ -141,33 +185,124 @@ def validate_email(uname):
-    projects = {}    
+    Returns a dictionary of projects keyed by project name,  and a list of rows that were not blank but failed
+    to parse correctly.
+    """    
```
extra spaces?

kamfonik commented:
Great, we will count Rob's approval of #16 as the second +1 and merge this on Monday.  I will probably tag you to update the form link then as well - we can check if the wordpress plugin is able to handle the multiple sections.
